### PR TITLE
package_facts: fix warning logic

### LINKED ISSUE #83502: package_facts fails silently when the manager is left to auto
### Summary

When investigating the issue #83501, I had a hard time understanding the issue, at first, because I left the package manager to `auto`, which effectively ends up making the conditions on lines [563](https://github.com/ansible/ansible/blob/7a3f9384cfa14ad9e2af23c75bc7e4a561a674ba/lib/ansible/modules/package_facts.py#L563) and [568](https://github.com/ansible/ansible/blob/7a3f9384cfa14ad9e2af23c75bc7e4a561a674ba/lib/ansible/modules/package_facts.py#L568) evaluates `apk` against `['auto']`, so the module emits no warning at all, in those conditions.

### Issue Type

Bug Report

### Component Name

package_facts

### Ansible Version

```console
ansible [core 2.17.1]
  config file = /usr/local/ansible/ansible.cfg
  configured module search path = ['/home/nobody/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/nobody/.local/share/pipx/venvs/ansible/lib/python3.12/site-packages/ansible
  ansible collection location = /home/nobody/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/nobody/.local/bin/ansible
  python version = 3.12.4 (main, Jun 27 2024, 00:10:56) [GCC 13.2.1 20240309] (/home/nobody/.local/share/pipx/venvs/ansible/bin/python)
  jinja version = 3.1.4
  libyaml = True
```

### Configuration

```console
CONFIG_FILE() = /usr/local/ansible/ansible.cfg
DEFAULT_HOST_LIST(/usr/local/ansible/ansible.cfg) = ['/usr/local/ansible/inventories']
DEFAULT_STDOUT_CALLBACK(/usr/local/ansible/ansible.cfg) = yaml

SHELL:
=====

sh:
__
remote_tmp(/usr/local/ansible/ansible.cfg) = /tmp
```

### OS / Environment

Alpine Linux 3.20.1

### Steps to Reproduce

Run the two tasks:
```yaml
- name: package_facts with manager set to apk 
  package_facts:
    manager: apk 

- debug:
    var: ansible_facts.packages
```
Then, compare the output to the one of 
```yaml
- name: package_facts with manager set to auto
  package_facts:
    manager: auto 

- debug:
    var: ansible_facts.packages
```

### Expected Results

I would have expected the two results to give the exact same warning, but having the manager set to `auto` totally silence the warning 

### Actual Results

```console
TASK [package_facts with manager set to apk] *************************************************
[WARNING]: Requested package manager apk was not usable by this module: Unable to list packages rc=0 : WARNING: opening from cache
https://dl-cdn.alpinelinux.org/alpine/v3.20/main: No such file or directory WARNING: opening from cache https://dl-
cdn.alpinelinux.org/alpine/v3.20/community: No such file or directory
ok: [localhost]

TASK [debug] *********************************************************************************
ok: [localhost] => 
  ansible_facts.packages: {}

TASK [package_facts with manager set to auto] ***********************************************
ok: [localhost]

TASK [debug] *********************************************************************************
ok: [localhost] => 
  ansible_facts.packages: {}
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by flowerysong**
##### SUMMARY

The logic had two nested `try` statements, but all of the exceptions were caught by the inner one so the outer `Failed to retrieve packages with %s` warning could never fire. This one should instead be the inner `try`, around just the single line that attempts to retrieve packages.

In order for this approach to work, the code it's calling needs to actually raise exceptions, which it did not.

Also, failing to list packages occurs after the package manager has been marked as found; this means that suppressing the warning when `manager=auto` can result in the list of packages being empty and no error or warning being visible. It's more reasonable to always warn in this situation.

Fixes #83502

##### ISSUE TYPE

- Bugfix Pull Request

### CODE REVIEW

**Review by bcoca — APPROVED** [review]

File: lib/ansible/modules/package_facts.py [hunk 1]
```diff
@@ -552,21 +552,18 @@ def main():
...
-                if pkgmgr in module.params['manager']:
-                    module.warn('Requested package manager %s was not usable by this module: %s' % (pkgmgr, to_text(e)))
-                continue
+                except Exception as e:
+                    module.warn('Failed to retrieve packages with %s: %s' % (pkgmgr, to_text(e)))
 
         except Exception as e:
             if pkgmgr in module.params['manager']:
-                module.warn('Failed to retrieve packages with %s: %s' % (pkgmgr, to_text(e)))
+                module.warn('Requested package manager %s was not usable by this module: %s' % (pkgmgr, to_text(e)))
```
Comment (Akasurde) [inline lib/ansible/modules/package_facts.py:521]: @flowerysong Could you please add a test for this change?

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/package_facts.py (hunk 1 above)
Comment (Akasurde) [reply lib/ansible/modules/package_facts.py:521]: rest LGTM

Comment (flowerysong) [reply lib/ansible/modules/package_facts.py:521]: ...but if I do that I might find out that things are broken and have to fix them. Bleh.

(The manager does not throw exceptions on init, nor indeed on `is_available()`.)

Comment (Akasurde) [reply lib/ansible/modules/package_facts.py:521]: At least we will cover the regression here.

Comment (webknjaz) [reply lib/ansible/modules/package_facts.py:521]: > ...but if I do that I might find out that things are broken and have to fix them.

If you don't, where would we have a proof that your patch …actually works?

**Review by s-hertel — APPROVED** [review]

**Review by Akasurde — APPROVED** [review]

File: changelogs/fragments/package_facts_warnings.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+  - package_facts - returns the correct warning when package listing fails.
+  - package_facts - no longer fails silently when the selected package manager is unable to list packages.
```
Comment (s-hertel) [inline changelogs/fragments/package_facts_warnings.yml:3]:
```suggestion
  - package_facts - no longer fails silently when the selected package manager is unable to list packages (https://github.com/ansible/ansible/issues/83502).
```

Comment (s-hertel) [reply changelogs/fragments/package_facts_warnings.yml:3]: Sorry, thought I'd link the issue in the porting guide quickly before merging but CI had other ideas. Closing/reopening to fix it :crossed_fingers:

Comment (s-hertel) [reply changelogs/fragments/package_facts_warnings.yml:3]: Reverted it

**Pull request merged**